File: app/src/main/java/me/zheteng/android/videosaver/saver/extractor/MetaVideoExtractor.java
package me.zheteng.android.videosaver.saver.extractor;

import android.support.annotation.Nullable;
import me.zheteng.android.videosaver.saver.Extractor;
import me.zheteng.android.videosaver.saver.ParserUtils;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 从Meta信息里抽取视频地址
 */
public class MetaVideoExtractor implements Extractor {
    @Override
    public @Nullable String extract(String html) {
        Pattern pattern = Pattern.compile("<meta([^>]+property=\"([^\"]*)\"[^>]+content=\"([^\"]*)\"|"
                        + "[^>]+content=\"([^\"]*)\"[^>]+property=\"([^\"]*)\")[^>]*>",
                Pattern.CASE_INSENSITIVE);
        Matcher matcher = pattern.matcher(html);
        Map<String, String> hits = new HashMap<>();
        while (matcher.find()) {
            if (matcher.group(2) != null) {
                hits.put(matcher.group(2), matcher.group(3));
            } else if (matcher.group(5) != null) {
                hits.put(matcher.group(5), matcher.group(4));
            }
        }

        String videoUrl = findVideoUrl(hits, "og:video:url");
        if (ParserUtils.isEmpty(videoUrl)) {
            videoUrl = findVideoUrl(hits, "og:video");
        }
        if (ParserUtils.isEmpty(videoUrl)) {
            videoUrl = findVideoUrl(hits, "og:video:secure_url");
        }
        return videoUrl;
    }

    @Nullable
    private String findVideoUrl(Map<String, String> hits, String key) {
        String videoUrl = null;
        for (Map.Entry<String, String> entry : hits.entrySet()) {
            if (entry.getKey().equalsIgnoreCase(key)) {
                videoUrl = entry.getValue();
                break;
            }
        }
        return videoUrl;
    }
}

File: app/src/main/kotlin/me/zheteng/android/videosaver/parse/extractor/MetaVideoExtractor.kt

package me.zheteng.android.videosaver.parse.extractor

import me.zheteng.android.videosaver.parse.Extractor
import me.zheteng.android.videosaver.parse.ParserUtils
import me.zheteng.android.videosaver.parse.Video
import java.util.*
import java.util.regex.Pattern


/**
 * 从Meta信息里抽取视频地址
 */
class MetaVideoExtractor : Extractor {
    override fun extract(html: String): List<Video> {
        val pattern = Pattern.compile("<meta([^>]+property=\"([^\"]*)\"[^>]+content=\"([^\"]*)\"|" + "[^>]+content=\"([^\"]*)\"[^>]+property=\"([^\"]*)\")[^>]*>",
                Pattern.CASE_INSENSITIVE)
        val matcher = pattern.matcher(html)
        val hits = HashMap<String, String>()
        while (matcher.find()) {
            if (matcher.group(2) != null) {
                hits.put(matcher.group(2), matcher.group(3))
            } else if (matcher.group(5) != null) {
                hits.put(matcher.group(5), matcher.group(4))
            }
        }

        var videoUrl: String? = findVideoUrl(hits, "og:video:url")
        if (ParserUtils.isEmpty(videoUrl)) {
            videoUrl = findVideoUrl(hits, "og:video")
        }
        if (ParserUtils.isEmpty(videoUrl)) {
            videoUrl = findVideoUrl(hits, "og:video:secure_url")
        }

        val list = ArrayList<Video>()
        if (!ParserUtils.isEmpty(videoUrl)) {
            list.add(Video(videoUrl!!))
        }

        return list

    }

    private fun findVideoUrl(hits: Map<String, String>, key: String): String? {
        var videoUrl: String? = null
        for ((key1, value) in hits) {
            if (key1.equals(key, ignoreCase = true)) {
                videoUrl = value
                break
            }
        }
        return videoUrl
    }
}